File: packages/core/src/canonical.ts
import { CanonicalJsonError } from "./errors.js";

/**
 * Deterministic JSON serialization (JCS-style): object keys sorted by UTF-16
 * code unit, no insignificant whitespace, `undefined` properties dropped.
 *
 * This is the byte-level contract behind receipt signatures and the §11
 * cross-language golden files, so it is deliberately strict:
 *
 * - non-integer and non-finite numbers are rejected (float formatting differs
 *   between languages — encode fractional values as decimal strings)
 * - bigints are rejected (encode as decimal strings, like receipt `amount`)
 * - functions, symbols and class instances with custom `toJSON` are rejected
 */
export function canonicalJson(value: unknown): string {
  return write(value, new WeakSet());
}

export function canonicalBytes(value: unknown): Uint8Array {
  return new TextEncoder().encode(canonicalJson(value));
}

function write(value: unknown, seen: WeakSet<object>): string {
  if (value === null) return "null";

  switch (typeof value) {
    case "string":
      return JSON.stringify(value);
    case "boolean":
      return value ? "true" : "false";
    case "number": {
      if (!Number.isFinite(value)) {
        throw new CanonicalJsonError(`non-finite number is not canonicalizable: ${value}`);
      }
      if (!Number.isInteger(value)) {
        throw new CanonicalJsonError(
          `non-integer number ${value} is not canonicalizable; use a decimal string`,
        );
      }
      // Number.isInteger(-0) is true and String(-0) is "0", so -0 normalizes.
      return String(value);
    }
    case "bigint":
      throw new CanonicalJsonError(
        `bigint ${value} is not canonicalizable; encode it as a decimal string`,
      );
    case "undefined":
      throw new CanonicalJsonError("undefined is not canonicalizable");
    case "object":
      break;
    default:
      throw new CanonicalJsonError(`${typeof value} is not canonicalizable`);
  }

  const obj = value as object;
  if (seen.has(obj)) throw new CanonicalJsonError("circular reference");
  seen.add(obj);
  try {
    if (Array.isArray(obj)) {
      const items = obj.map((item) => write(item === undefined ? null : item, seen));
      return `[${items.join(",")}]`;
    }
    if (Object.getPrototypeOf(obj) !== Object.prototype && Object.getPrototypeOf(obj) !== null) {
      throw new CanonicalJsonError(
        `only plain objects are canonicalizable, got ${obj.constructor?.name ?? "unknown"}`,
      );
    }
    const record = obj as Record<string, unknown>;
    // Default sort compares UTF-16 code units, which is what JCS specifies.
    const keys = Object.keys(record)
      .filter((key) => record[key] !== undefined)
      .sort();
    const entries = keys.map((key) => `${JSON.stringify(key)}:${write(record[key], seen)}`);
    return `{${entries.join(",")}}`;
  } finally {
    seen.delete(obj);
  }
}

File: packages/python/src/octroi/canonical.py
"""Deterministic JSON serialization.

This is the byte-level contract behind receipt signatures and the §11
cross-language golden files, so it mirrors ``canonical.ts`` exactly: object
keys sorted, no insignificant whitespace, ``None``-valued keys kept, missing
keys dropped, and anything whose textual form differs between languages
refused outright.

The one documented divergence: JavaScript sorts keys by UTF-16 code unit,
Python by Unicode code point. These agree for every character below U+10000,
and differ only for astral-plane keys (emoji, rare CJK). Octroi never emits
such keys; if you add one, the golden files will catch it.
"""

from __future__ import annotations

import json
from typing import Any, Set

__all__ = ["CanonicalJsonError", "canonical_json", "canonical_bytes"]


class CanonicalJsonError(ValueError):
    """A value cannot be canonicalized byte-identically across languages."""

    code = "canonical_json"


def canonical_json(value: Any) -> str:
    return _write(value, set())


def canonical_bytes(value: Any) -> bytes:
    return canonical_json(value).encode("utf-8")


def _write(value: Any, seen: Set[int]) -> str:
    if value is None:
        return "null"

    # bool before int: in Python, True is an int, and "1" is not "true".
    if isinstance(value, bool):
        return "true" if value else "false"

    if isinstance(value, str):
        # ensure_ascii=False keeps non-ASCII literal, matching JSON.stringify;
        # control characters still take the \u form in both languages.
        return json.dumps(value, ensure_ascii=False)

    if isinstance(value, int):
        return str(value)

    if isinstance(value, float):
        # Float formatting differs between languages, so fractional values are
        # decimal strings — which is why receipt `amount` is "4000".
        raise CanonicalJsonError(
            f"non-integer number {value!r} is not canonicalizable; use a decimal string"
        )

    if isinstance(value, (list, tuple)):
        if id(value) in seen:
            raise CanonicalJsonError("circular reference")
        seen.add(id(value))
        try:
            return "[" + ",".join(_write(item, seen) for item in value) + "]"
        finally:
            seen.discard(id(value))

    if isinstance(value, dict):
        if id(value) in seen:
            raise CanonicalJsonError("circular reference")
        seen.add(id(value))
        try:
            for key in value.keys():
                if not isinstance(key, str):
                    # Checked before sorted(): mixed-type keys make sorted()
                    # raise a bare TypeError, which is not this module's error.
                    raise CanonicalJsonError(
                        f"object keys must be strings, got {type(key).__name__}"
                    )
            items = []
            for key in sorted(value.keys()):
                item = value[key]
                # A key explicitly set to None survives as null (JS keeps null
                # too); only absent keys are absent.
                items.append(f"{json.dumps(key, ensure_ascii=False)}:{_write(item, seen)}")
            return "{" + ",".join(items) + "}"
        finally:
            seen.discard(id(value))

    raise CanonicalJsonError(f"{type(value).__name__} is not canonicalizable")
